object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java

```diff
@@ -111,14 +116,143 @@ void checkMustCallInvoked(ControlFlowGraph cfg) {
+    if (assignmentContext != null) {
+      Element elementForType = assignmentContext.getElementForType();
+      if (assignmentContext instanceof AssignmentLhsContext) {
+        // lhs should be a local variable
+        assignedToOwning = isOwningAssignmentLhs(elementForType);
+      } else if (assignmentContext instanceof MethodParameterContext) {
+        // must be an @Owning parameter
+        assignedToOwning = typeFactory.getDeclAnnotation(elementForType, Owning.class) != null;
+      } else if (assignmentContext instanceof MethodReturnContext) {
+        // must be an @Owning return
+        assignedToOwning =
+            TRANSFER_OWNERSHIP_AT_RETURN
+                ? typeFactory.getDeclAnnotation(
+                        TreeUtils.elementFromTree(assignmentContext.getContextTree()),
+                        NotOwning.class)
+                    == null
+                : typeFactory.getDeclAnnotation(elementForType, Owning.class) != null;
+      } else if (assignmentContext instanceof LambdaReturnContext) {
+        // TODO handle this case.  For now we will report an error
+      } else {
+        throw new BugInCF("unexpected AssignmentContext type " + assignmentContext.getClass());
+      }
+    }
+    if (!assignedToOwning) {
+      // check if @CalledMethods type of return satisfies the @MustCall obligation
+      AnnotationMirror cmAnno =
+          typeFactory.getAnnotatedType(tree).getAnnotationInHierarchy(typeFactory.top);
+      if (!calledMethodsSatisfyMustCall(mustCallVal, cmAnno)) {
+        checker.reportError(
+            tree,
+            "required.method.not.called",
+            MustCallInvokedChecker.formatMissingMustCallMethods(mustCallVal),
+            TreeUtils.typeOf(tree).toString(),
+            "never assigned to an @Owning location");
+      }
+    }
+  }
+
+  /**
+   * Does an element represents an assignment left-hand side that can take ownership?
+   *
+   * @param elem the element
+   * @return {@code true} iff {@code elem} represents a local variable, a try-with-resources
+   *     variable, or an {@code @Owning} field
+   */
+  private boolean isOwningAssignmentLhs(Element elem) {
+    return elem != null
+        && (elem.getKind().equals(ElementKind.LOCAL_VARIABLE)
+            || elem.getKind().equals(ElementKind.RESOURCE_VARIABLE)
+            || (elem.getKind().equals(ElementKind.FIELD)
+                && typeFactory.getDeclAnnotation(elem, Owning.class) != null));
+  }
+
+  /**
+   * Checks for cases where we do not need to ensure that a method invocation gets pseudo-assigned
+   * to a variable / field that takes ownership. We can skip the check when the invoked method's
+   * return type is {@link org.checkerframework.common.returnsreceiver.qual.This}, the invocation is
+   * a super constructor call, or the method's return type is annotated {@link NotOwning}
+   */
+  private boolean shouldSkipInvokePseudoAssignCheck(Tree callTree) {
+    if (callTree instanceof MethodInvocationTree) {
```

Comment:
I think it's preferred to compare `TreeKind`s rather than do `instanceof` on `Tree`s. So this check would become `callTree.getKind == TreeKind.METHOD_INVOCATION`.

Reply:
I don't remember why, though.